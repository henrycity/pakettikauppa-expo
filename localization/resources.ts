// the translations
// (tip move them in a JSON file and import them)
const resources = {
  en: {
    translation: {
      profile: 'Profile',
      shipments: 'Shipments',
      settings: 'Settings',
    },
  },
  fi: {
    translation: {
      profile: 'Profiili',
      shipments: 'Lähetykset',
      settings: 'Asetukset',
    },
  },
  se: {
    translation: {
      profile: 'Profil',
      shipments: 'Transporter',
      settings: 'Inställningar',
    },
  },
}

export default resources

export const languages = Object.keys(resources)
