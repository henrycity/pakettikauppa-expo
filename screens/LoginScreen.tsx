import * as Google from 'expo-auth-session/providers/google'
import * as SecureStore from 'expo-secure-store'
import * as WebBrowser from 'expo-web-browser'
import * as React from 'react'
import { StyleSheet, Button, Platform } from 'react-native'

import { Text, View } from '../components/Themed'
import useLogin from '../hooks/useLogin'

WebBrowser.maybeCompleteAuthSession()

export default function LoginScreen(): JSX.Element {
  const { login, disabled } = useLogin()

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Login or register with a Google account</Text>
      <View
        style={styles.separator}
        lightColor="#eee"
        darkColor="rgba(255,255,255,0.1)"
      />

      <Button
        disabled={disabled}
        title="Login"
        onPress={() => {
          login()
        }}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
})
