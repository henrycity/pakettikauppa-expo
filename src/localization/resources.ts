// the translations
// (tip move them in a JSON file and import them (which should make the app load faster when there are lots of resources))
const resources = {
  en: {
    translation: {
      profile: 'Profile',
      reports: 'Reports',
      statistics: 'Statistics',
      shipments: 'Shipments',
      settings: 'Settings',
      logout: 'Log out',
      login: 'Login',
    },
  },
  fi: {
    translation: {
      profile: 'Profiili',
      reports: 'Raportit',
      statistics: 'Tilastot',
      shipments: 'Lähetykset',
      settings: 'Asetukset',
      logout: 'Kirjaudu ulos',
    },
  },
  se: {
    translation: {
      profile: 'Profil',
      reports: 'Raporter',
      statistics: 'Statistik',
      shipments: 'Transporter',
      settings: 'Inställningar',
      logout: 'Logga ut',
    },
  },
}

export default resources

export const languages = Object.keys(resources)
