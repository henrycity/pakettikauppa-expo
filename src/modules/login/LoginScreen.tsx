import * as WebBrowser from 'expo-web-browser'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { TouchableOpacity } from 'react-native'

import Styles from '../../common/Styles'
import { useThemedColors, View, Text } from '../../common/Themed'
import RegistrationModal from './components/RegistrationModal'
import useLogin from './hooks/useLogin'

WebBrowser.maybeCompleteAuthSession()

export default function LoginScreen(): JSX.Element {
  const { login } = useLogin()
  const { t } = useTranslation()
  const clickHandler = () => {
    login()
  }

  const theme = useThemedColors()
  return (
    <View style={Styles.container}>
      <Text style={Styles.title}>Login or register with a Google account</Text>
      <View
        style={Styles.separator}
        lightColor="#eee"
        darkColor="rgba(255,255,255,0.1)"
      />
      <TouchableOpacity
        onPress={clickHandler}
        style={[Styles.normalButton, { backgroundColor: theme.buttonColor }]}
        accessibilityLabel="Login button"
      >
        <Text style={[Styles.buttonLabel, { color: theme.buttonTextColor }]}>
          {t('LOGIN')}
        </Text>
      </TouchableOpacity>
      <View style={Styles.gap} />

      <RegistrationModal />
    </View>
  )
}
