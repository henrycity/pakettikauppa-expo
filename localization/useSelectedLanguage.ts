import { useState, useEffect, Dispatch, SetStateAction } from "react"
import AsyncStorage from "@react-native-async-storage/async-storage"
import * as Localization from "expo-localization"
import { languages } from './resources'

type useSelectedLanguageReturnType = [string, Dispatch<SetStateAction<string>>]

export default function useSelectedLanguage(): useSelectedLanguageReturnType {
  const [selected, setSelected] = useState('')
  useEffect(() => {
    getSelectedLanguage()
      .then(lng => setSelected(lng))
  }, [])

  return [selected, setSelected]
}

async function getSelectedLanguage() {
  let selected = await getStoredLanguage()
  if (!selected) {
    const locale = Localization.locale
    if (languages.includes(locale))
      selected = locale
    else
      selected = 'en'
  }
  return selected
}

async function getStoredLanguage(): Promise<string | null> {
  return AsyncStorage.getItem('lng')
}
