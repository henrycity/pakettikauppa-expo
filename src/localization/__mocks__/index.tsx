import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

import resources from '../resources'

export { languages } from '../resources'

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const useSelectedLanguage = () => ['en', (_selected: string) => {}]

export async function initializeLocalization(): Promise<void> {
  i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
      resources,
      lng: 'en',
      fallbackLng: 'en',

      keySeparator: false,

      interpolation: {
        escapeValue: false, // react already safes from xss
      },
    })
}
