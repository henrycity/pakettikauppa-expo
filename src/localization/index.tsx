import AsyncStorage from '@react-native-async-storage/async-storage'
import * as Localization from 'expo-localization'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

import resources from './resources'

export { languages } from './resources'
export { default as useSelectedLanguage } from './hooks/useSelectedLanguage'

export async function initializeLocalization(): Promise<void> {
  const language = await AsyncStorage.getItem('lng')

  i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
      resources,
      lng: language ?? Localization.locale,
      fallbackLng: 'en',

      keySeparator: false,

      interpolation: {
        escapeValue: false, // react already safes from xss
      },
    })

  setLanguageChangedEventListener()
}

function setLanguageChangedEventListener(): void {
  i18n.on('languageChanged', (lng) => {
    if (lng) {
      // Split because Localization.locale format is "en-EN" instead of "en"
      storeLanguage(lng.split('-')[0])
    }
  })
}

function storeLanguage(language: string): Promise<void> {
  return AsyncStorage.setItem('lng', language)
}
