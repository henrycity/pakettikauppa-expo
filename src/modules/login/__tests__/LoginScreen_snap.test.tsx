import { render, fireEvent, cleanup } from '@testing-library/react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import { initializeLocalization } from '../../../localization'
import LoginScreen from '../LoginScreen'

it('LoginScreen renders correctly', () => {
  initializeLocalization()
  const tree = renderer.create(<LoginScreen />).toJSON()

  expect(tree).toMatchSnapshot()
})
