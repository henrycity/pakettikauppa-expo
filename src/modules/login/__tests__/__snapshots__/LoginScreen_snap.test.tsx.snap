// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#233385",
        },
        Object {
          "fontFamily": "Muli",
        },
        Array [
          Object {
            "fontFamily": "Rubik-Bold",
            "fontSize": 26,
          },
          Object {
            "textAlign": "center",
          },
        ],
      ]
    }
  >
    Log in or register with a Google account
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
        },
        Object {
          "height": 1,
          "marginVertical": 30,
          "width": "80%",
        },
      ]
    }
  />
  <View>
    <View
      accessibilityLabel="Login button"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ff2e33",
          "borderRadius": 100,
          "justifyContent": "center",
          "minWidth": 85,
          "opacity": 1,
        }
      }
      testID="Log in"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#233385",
            },
            Object {
              "fontFamily": "Muli",
            },
            Array [
              Object {
                "fontFamily": "Rubik-Bold",
                "fontSize": 14,
                "padding": 10,
                "textTransform": "uppercase",
              },
              Object {
                "color": "#fff",
                "textDecorationLine": "none",
              },
              undefined,
            ],
          ]
        }
      >
        Log in
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "height": 10,
        },
      ]
    }
  />
  <View>
    <View
      accessibilityLabel="Registration button"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ff2e33",
          "borderRadius": 100,
          "justifyContent": "center",
          "minWidth": 85,
          "opacity": 1,
        }
      }
      testID="register"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#233385",
            },
            Object {
              "fontFamily": "Muli",
            },
            Array [
              Object {
                "fontFamily": "Rubik-Bold",
                "fontSize": 14,
                "padding": 10,
                "textTransform": "uppercase",
              },
              Object {
                "color": "#fff",
                "textDecorationLine": "none",
              },
              undefined,
            ],
          ]
        }
      >
        Register
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "height": 10,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        undefined,
      ]
    }
  />
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="modalTest"
    visible={false}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#233385",
            },
            Object {
              "fontFamily": "Muli",
            },
            Object {
              "fontFamily": "Rubik-Bold",
              "fontSize": 26,
            },
          ]
        }
      >
        Please enter your details
      </Text>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "height": 10,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            undefined,
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#233385",
            },
            Object {
              "fontFamily": "Muli",
            },
            undefined,
          ]
        }
      >
        Email:
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        defaultValue=""
        onChangeText={[Function]}
        placeholder="Email:"
        placeholderTextColor="#8cb1b7"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#8cb1b7",
            },
            Object {
              "fontFamily": "Muli",
            },
            Array [
              Object {
                "borderBottomWidth": 1,
                "fontFamily": "Muli",
                "padding": 3,
                "width": 200,
              },
              Object {
                "color": "#233385",
              },
              Object {
                "borderColor": "#8cb1b7",
              },
            ],
          ]
        }
        testID="Email"
        underlineColorAndroid="transparent"
      />
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "height": 10,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#233385",
            },
            Object {
              "fontFamily": "Muli",
            },
            undefined,
          ]
        }
      >
        Vat ID:
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        defaultValue=""
        onChangeText={[Function]}
        placeholder="Vat ID:"
        placeholderTextColor="#8cb1b7"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#8cb1b7",
            },
            Object {
              "fontFamily": "Muli",
            },
            Array [
              Object {
                "borderBottomWidth": 1,
                "fontFamily": "Muli",
                "padding": 3,
                "width": 200,
              },
              Object {
                "color": "#233385",
              },
              Object {
                "borderColor": "#8cb1b7",
              },
            ],
          ]
        }
        testID="VAT"
        underlineColorAndroid="transparent"
      />
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "height": 10,
            },
          ]
        }
      />
      <View>
        <View
          accessibilityLabel="Submit registration"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ff2e33",
              "borderRadius": 100,
              "justifyContent": "center",
              "minWidth": 85,
              "opacity": 1,
            }
          }
          testID="submit"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#233385",
                },
                Object {
                  "fontFamily": "Muli",
                },
                Array [
                  Object {
                    "fontFamily": "Rubik-Bold",
                    "fontSize": 14,
                    "padding": 10,
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#fff",
                    "textDecorationLine": "none",
                  },
                  undefined,
                ],
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "height": 10,
            },
          ]
        }
      />
      <View>
        <View
          accessibilityLabel="Close registration modal"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ff2e33",
              "borderRadius": 100,
              "justifyContent": "center",
              "minWidth": 85,
              "opacity": 1,
            }
          }
          testID="close"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#233385",
                },
                Object {
                  "fontFamily": "Muli",
                },
                Array [
                  Object {
                    "fontFamily": "Rubik-Bold",
                    "fontSize": 14,
                    "padding": 10,
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#fff",
                    "textDecorationLine": "none",
                  },
                  undefined,
                ],
              ]
            }
          >
            Close
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
